### ISSUE TITLE
mutate: add customizable filter for unproductive mutants

### ISSUE THREAD

**Issue by christoffer-nylen** [joakim-brannstrom/dextool#917]
Assume:
* There is a file `a.c` with 534 LoC.
* The file contains 37 invokations of a function `addLoggMessage(param, param)` that is used for debugging purposes, and there are 359 similar cases in the full project.
* The `addLoggMessage` usage (e.g., invokation order and contents of argument) is highly coupled to the `a.c` implementation.
* The correctness of the `addLoggMessage()` usage is not critical.

Scenario:
* The user would like to focus on critical/productive test scenarios.

Problem:
* In `a.c`, 37 invokations of `addLoggMessage(param, param)` gets reported as survived.
* The reported statements draw attention from critical problems.

Suggestion:
* Add some kind of filtering function so that the user can specify to ignore all invokations of `addLoggMessage()`.

**Comment by joakim-brannstrom** [comment]
Yes this is coupled to how to improve NOMUT. Suppressing mutants.
I have had the idea to let the user specify uninteresting functions.

**Comment by joakim-brannstrom** [comment]
For now what exist is to mark those lines with NOMUT

**Comment by peatersson** [comment]
The idea I've started implementing, but put on hold, was that the user specified mutantID. 
Example:
`dextool mutate admin --suppress 52`

I realized that this quickly became problematic in terms of scale. A user cant be expected to write that commando for every single mutant, if there are 359 as in your example. 

I know I wouldn't.

Are we maybe thinking of an addition to the config-file? Something like:
```
# Specify which functions you would like to exclude from the mutation testing
# Note: this will affect all kinds of mutants surrounding those functions.
# For example: SDL won't remove the call to the function, or the function-body.
suppress-functions = ["addLoggMessage", "print", "randomFunctionToSuppress"]
```
And then we add the ability to run admin-mode and suppress those now added to the config?

**Comment by joakim-brannstrom** [comment]
> The idea I've started implementing, but put on hold, was that the user specified mutantID.

Did you consider suppressing a whole line?

> Example:
> `dextool mutate admin --suppress 52`
> 
> I realized that this quickly became problematic in terms of scale. A user cant be expected to write that commando for every single mutant, if there are 359 as in your example.

I agree. Another problem is that they constantly change. The checksum is *more* stable.

> Are we maybe thinking of an addition to the config-file? Something like:
> 
> ```
> # Specify which functions you would like to exclude from the mutation testing
> # Note: this will affect all kinds of mutants surrounding those functions.
> # For example: SDL won't remove the call to the function, or the function-body.
> suppress-functions = ["addLoggMessage", "print", "randomFunctionToSuppress"]
> ```

Yes, in that direction. But probably have to allow the user to be more precis?
This is why I think it is *hard* to do. How to specify the mutants in a sane way that the user wont cry rivers over?
A line number is kind a easy. A bit brute force but doable.

> And then we add the ability to run admin-mode and suppress those now added to the config?

Shouldn't need to run in admin-mode. Being in the config file should be enough.

**Comment by peatersson** [comment]
> Did you consider suppressing a whole line?

Yes, I started with that. But I remembered (maybe wrongly) that the user who requested this said that maybe not ALL mutants on a row needed to be suppressed. I therefore went with the mutantID-solution first. But as I said, it was put on hold early in the phase.

> I agree. Another problem is that they constantly change. The checksum is more stable.

Oh definitely. 

@christoffer-nylen Extremely good edits on the issue! :)

**Comment by peatersson** [comment]
User feedback:
_Working with inspected code makes it difficult to use current NOMUT. The use of marking mutants by ID in a command is a valid option._

I am thinking that I can start with implementing this, so that it is atleast possible to do something. We can then evaluate and improve it with more options in the config-file, for example marking certain function-calls.

**Comment by joakim-brannstrom** [comment]
I discussed this with a user. How to use this to mark a mutant as unproductive in a pull request.
The simplest way we could come up with is to use the configuration file in the PR to mark it in the database.
If the approach we are talking about here makes that possible then we have another tool for our box to realize "CI mode".

**Comment by joakim-brannstrom** [comment]
#946 changes how mutants are calculated which is one outcome of the discussions on this issue.

**Comment by peatersson** [comment]
#953 intend to solve parts of this problem by allowing the user to mark mutants manually by id via an admin command. There is obvious issues with this approach, one of them being that MutationId is very unstable and changes when the analyze phase is run again.

This was however a request from a user, so atleast providing the ability to do this increases short term usability. From here, we can work towards a more complete solution or alternative (both can still be available in the tool).
